GHC's base libraries: Combining stability with innovation (#51)
haskellfoundation/tech-proposals: proposals/accepted/051-ghc-base-libraries.rst

================
GHC's base libraries
================
----------
Combining stability with innovation
----------

:Date: June 2023
:Authors:
  John Ericson — secretary,
  Ben Gamari — GHC,
  Adam Gundry — GHC,
  Andrew Lelechenko — CLC,
  Julian Ospald — CLC,
  Simon Peyton Jones — GHC

.. sectnum::
.. contents::

Introduction
=========

> This document describes an plan agreed between the GHC Team and the Core Libraries Committee, saying how we plan to work in partnership to reconcile the goals of innovation and stability for GHC and its ecosystem.

Reviewer: ```suggestion
This document describes a plan agreed between the GHC Team and the Core Libraries Committee, saying how we plan to work in partnership to reconcile the goals of innovation and stability for GHC and its ecosystem.
```

Goals
=====

This proposal seeks to reconcile two goals, both of which are complex and multi-faceted.

1. **The Core Libraries Committee seeks to exercise its control over decisions affecting 'base' (API, performance, semantics), including its dependencies.**
   Managing API changes of ``base`` is the CLC's primary mandate.
   If ``base`` changes gratuitously, everyone suffers.

2. **The GHC team seeks the freedom to:**

   - Innovate in the language design.
     GHC has hundreds of extensions, and people suggest more all the time, via the GHC Proposals process.

   - Move rapidly to fix bugs, improve performance, and refactor GHC's internals to pay down technical debt.

The two goals are in risk of conflict, and we have seen such conflict in the past.
The purpose of this proposal is to reconcile these goals, in a way that satisfies all parties.

Each of us places different emphasis on these goals, but respects them both.
The CLC does not want to constrain GHC unnecessarily; and the GHC team does not want to undermine CLC's efforts, indeed quite the reverse!

> The proposal is based on `HF Proposal: standard library reform <https://github.com/haskellfoundation/tech-proposals/pull/47>`__, but is independent of it.

Reviewer 2: Add: 

You can find more background in [CLC issue #146](https://github.com/haskell/core-libraries-committee/issues/146).

Author: Like this?
```suggestion
The proposal is based on `HF Proposal: standard library reform <https://github.com/haskellfoundation/tech-proposals/pull/47>`__, but is independent of it.
You can find more background in `CLC issue #146 <https://github.com/haskell/core-libraries-committee/issues/146>`__.
```

Reviewer 2: Yes, that's what I was thinking.  Just to keep pointers to history.

   Beyond that, it has no interest in the implementation details (e.g. alpha renaming, moving things between modules, comments).

5. In curating the ``base`` API, it is immaterial where code lives.
   For example, if a change to ``ghc-prim`` changes the ``base`` API (as defined above) the GHC developers must consult the CLC.
   The fact that the change isn't physically part of the ``base`` package is immaterial.
   (Incidentally, ``base`` and ``ghc-prim`` are both part of the same GitHub repository, which also includes GHC's source code.)

Proposal
========

We propose to divide ``base`` into three packages:

- ``ghc-internals``: exposes aspects of GHC's internals that may be of interest to "hard-core" developers interested in maximum performance (see `Nikita's blog post <https://nikita-volkov.github.io/internal-convention-is-a-mistake/>`__).
  The API of ``ghc-internals`` is fully under the control of the GHC team, and of no direct interest to the CLC — only its effects on the API of base.

- ``base``: as now, whose API is curated by CLC.
  Depends on ``ghc-internals``, and hence on ``ghc-bignum`` and ``ghc-prim``.

> - ``ghc-experimental``, initially empty, depends on ``base`` and on ``ghc-internals``.

Reviewer 3: OK, so the governance structure is:

- `base` governed by CLC and CLC proposals
- `ghc-internals` governed solely by GHC team, no proposal necessary
- `ghc-experimental` governed by GHC proposal process

The Win here is that GHC Team can unilaterally make changes to `ghc-internals` without asking anyone, which allows them to iterate quickly. GHC Proposals can go in `ghc-experimental` without involving CLC directly. And CLC has less extraneous work. Feels like a win-win-win to me.

>   Functions and data types here are intended to have their ultimate home in base, but while they are settling down they are subject to much weaker stability guarantees.

Reviewer 3: Something I'm always wary of here is the tendency for things to go stale - not quite popular enough to move to `base`, but useful enough to stay in `ghc-{experimental,internal}`. Is there room for a policy on how long something should be "unstable" before getting shelved or promoted?

Reviewer 4: 100% agree that `ghc-experimental` should have a clear lifecycle, and ideally an "up-or-out" policy of some kind.

  Generally, new functions and types introduced in GHC Proposals would start their life here.
  Example: new type families and type constructors for tuples, `GHC Proposal #475 <https://github.com/ghc-proposals/ghc-proposals/pull/475>`__.

  Another example: future APIs to access RTS statistics, which are fairly stable and user-exposed, but which are (by design) coupled closely to GHC's runtime and hence may change.

  As its name suggests, the API of ``ghc-experimental`` is curated by the GHC team, although the CLC is willing to offer (non-binding) opinions, if consulted.

All three packages conform rigorously to the PVP.
(But see Section 5.3)

Some observations about this structure:

- We should use all possible social and technical means to discourage people from depending directly on ``ghc-internals``, because if such dependencies become frequent and ossified, it will lead to future pain when the API changes.

>   Saying "we told you not to rely on it" may be true, but won't lessen that pain.

Reviewer 3: I also don't feel like this part of the proposal is well-motivated - specifically why `ghc-internals` should be treated separately from a potential `text-internals` or `bytestring-internals`. 

From my perspective, writing `build-depends: ghc-internals` is just as prone to pain as any other unbounded dependency. And `build-depends: ghc-internals == 9.6.4.*` should be perfectly safe, unless PVP is violated. But the *point* of this proposal is that we can make "breaking changes" to `ghc-internals` as a major version bump without incurring a major version bump for `base` - in other words, so we *can* be PVP compliant.

>   Saying "we told you not to rely on it" may be true, but won't lessen that pain.

Reviewer 2: > I also don't feel like this part of the proposal is well-motivated - specifically why ghc-internals should be treated separately from a potential text-internals or bytestring-internals.

It's really a response to a legitimate CLC concern that if it's too easy to depend on `ghc-internals` then that's what people will do.  And if that happens we might grow an ecosystem of libraries many of whch depend, perhaps by accident, on `ghc-internals`.  So every new release of GHC would force updates to all of those libraries.  You could say "if you depend on `ghc-internals` then that's what you get, but the pain is real and not mitigated by saying "I told you so".  To put it another way, it risks entirely bypassing the stability efforts of the CLC.

So that is, I think, the reason for this text. You are right that all the same issues apply to any `-internals` package, and one might wonder about mechanisms to discourage depending on them. But to avoid scope creep in this document we just stuck to `ghc-internals`.

>   Saying "we told you not to rely on it" may be true, but won't lessen that pain.

Reviewer 5: Discouragement is well motivated. Prohibition isn't.

We have to educate users more generally about the pattern we're trying to introduce here and how it relates to PVP, API stability, etc.

I could imagine that we try to summarize the result of these efforts for end users and library maintainers in a more distilled way and explain the pattern there in more depth.

I don't think we have to be particularly scared about ecosystem degrading.

1. Hackage has a reverse deps feature and we can monitor packages depending on ghc-internals easily.
2. There are many eyes on popular packages like aeson, servant, conduit etc. and I'm sure someone will scold any attempt at introducing a ghc-internals dependency
3. Even if a package accidentally does so, it's not irreversible. We can scold the maintainer and ask why they saw the need to do so.
4. "Pick your libraries carefully" already applies. Well educated maintainers who care about the ecosystem will play nicely.

>   Saying "we told you not to rely on it" may be true, but won't lessen that pain.

Reviewer 3: @simonpj I think that makes good sense. I want GHC to be able to innovate freely and iterate quickly, and I wouldn't want CLC to start encroaching on the `ghc-internals` package for this reason. Where I think I'd draw the line is at describing the problem, noting it is a known ecosystem-wide problem that deserves a good solution, and somehow put a reminder on the package to *use* that solution when it is implemented.

There are a few perspectives here, and I think it'd be good to be really clear about which perspective a bullet is trying to satisfy. This one appears to be trying to satisfy both "industrial users want fewer breaking changes when upgrading major versions of GHC" and "CLC wants to be able to promote stability for the ecosystem" - which are overlapping but subtly different concerns.

As an industry user and OSS contributor, the primary pain point of upgrading GHC is that there are breaking changes that require a large chunk of the ecosystem to be touched. `base` is usually a relatively small part of this pain, with `ghc-prim` and `template-haskell` causing far more issues. I don't expect that `ghc-internals` is going to meaningfully impact the work that gets done here.

If `ghc-internals` is banned by `cabal-install`, then I don't see how it is different from un-exposed modules in `base`. If `ghc-internals` is banned on Hackage, then industrial users will need to vendor code in order to share libraries that depend on it - which *increases* the pain of a major upgrade by requiring *applications* to depend on `ghc-internals` directly, rather than sharing a library which can abstract over multiple versions of `ghc-internals` via CPP or other techniques. If there are warnings, they will be ignored for good and bad reasons, and maintainers will have another package to consider when doing ecosystem upgrades, but the overall workload doesn't change much.

As a CLC member, `ghc-internals` makes my life easier by having fewer changes to `base`, and changes to `base` that can be backed up by real ecosystem use in `ghc-internals`. But if using `ghc-internals` is heavily discouraged, then it's difficult to identify whether or not something in that package is satisfying the needs of the ecosystem.

Summarizing,

- I don't think we should use strong wording like "all possible social and technical means." 
- I'm not even sure we need to discourage beyond calling the package `ghc-internals` and writing docs that it'll be a major version bump with breaking changes every GHC release.
- I *definitely* think that proposing specific technical means for discouraging folks is out-of-scope here.

>   Saying "we told you not to rely on it" may be true, but won't lessen that pain.

Reviewer 6: The more I think about it, the more I am convinced that "all possible social and technical means" is too strong. It is obviously not literal. We are not going to hire a crack commando unit who survive as soldiers of fortune to go to your house if you depend on it.

Better language might be "develop both social and technical mechanisms to discourage..."

  What mechanisms could we use?

  - The name ``ghc-internals`` is a pretty strong signal all by itself.

  - Cabal description and README explains how it is intended used (and not used).

  - Hoogle could (by default anyway) never show stuff from ``ghc-internals``.

  - Do not upload Haddocks for ``ghc-internals`` to Hackage.
    (Ditto ``ghc-prim``.) Need to make sure that if someone wants to follow the Haddock source-code link to (say) Functor, they should still find it regardless of where it is actually defined.

  - We could consider issuing a warning if you say ``-package ghc-internals`` (or ``ghc-bignum`` or ``ghc-prim``), one that was hard to silence.
    Since we can have module-level ``WARNING`` pragmas with custom categories, one way to realise this would be to pick a category and add such pragmas to every module in the relevant packages, though we might want to do something more systematic.
    The text of the warnings could encourage users to

    - switch to a function exposed by base, and/or
    - petition the CLC to expose this super-useful function from base.

  - ``cabal check`` (a per-package check) could warn on packages that use ``ghc-internals``.

>   - ...what else?

Reviewer 7: > We should use all possible social and technical means to discourage people from depending directly on ``ghc-internals``

Really all? Including GHC/Cabal simply refusing to allow packages called  `ghc-internals` if the package currently built isn't `base` or another internal whitelisted package?

But it says “discourage”, not “prevent”, which sounds reasonable.

Maybe more realistic: hackage could reject uploads of packages that depend on `ghc-internals`?

Author: > But it says “discourage”, not “prevent”, which sounds reasonable.

Yes that is key. If people want to tinker with the internals of some open source software, we should not stop them. But we should make sure they are very aware they are stepping outside the stable safe zone. And likewise we ought to someday make anyone that uses the tinker's software (transitively!) also aware they are exposed to instability.

Author: (The difference between `ghc-prim` and `ghc-internal` also isn't clear. But one answer is

> GHC devs can structure (and restructure) their internals however they like, including dividing definitions *arbitrarily* between `ghc-prim` and `ghc-internals` if they so desire, so this question doesn't need to be answered in this document.

Another answer is

> `ghc-prim` is (mostly, perhaps should be entirely) auto-generated docs for code that doesn't actually exist in textual form (it's just hard-coded in GHC), whereas `ghc-internals` is human-written code.

)

  The existence of ``ghc-experimental`` should substantially ameliorate the difficulty that many GHC Proposals have a library-function component, but it is unlikely to be a *stable* API (having just been invented) and is therefore in conflict with the CLC's goals.

  As they become stable, the CLC may want to consider adopting the new types and functions from ``ghc-experimental`` into ``base``.
  (But CLC would not expect to curate the API of ``ghc-experimental``.)

- Perhaps ``ghc-experimental`` should be in the purview of the GHC Proposals process.
  GHC devs should not just make up random APIs and pop them into ``ghc-experimental``; a scrutiny process would be valuable.

- Under this proposal, there is initially no change (whatsoever) to the API exposed by ``base``, or its performance characteristics.
  The impact on clients should therefore be zero.

  Over time, the GHC developers may make CLC proposals to remove types and functions that are currently in the ``base`` API, but are in truth part of GHC's implementation, and were originally exposed by historical accident.

>   But these are *future*\ proposals.

Reviewer 8: ```suggestion
  But these are *future* proposals.
```

>   To make the transition suggested in these future proposals easier to manage, we have in progress a "deprecated exports" mechanism that will ease such transitions.

Reviewer 8: ```suggestion
  To make the transition suggested in these future proposals easier to manage, we have in progress a ["deprecated exports"](https://github.com/ghc-proposals/ghc-proposals/pull/595) mechanism that will ease such transitions.
```

  For a transitional period, ``base`` can continue to export the function, but with a deprecation warning saying something like:

    This is going to disappear from base.
    You probably don't want to use it at all.

>     But if you absolutely must, get it from ``ghc-internals``.

Reviewer 8: We can't say in the same document that we should use "all means necessary" to avoid `ghc-internals` dependencies, and also recommend people to import a function from there.

```suggestion
    If you absolutely must, define it yourself locally, or if this is impossible, let the CLC and GHC team know!
```

Author: I would expect `ghc-internals` to mainly contain things one cannot define locally? I would expect also some zany integrate-with-ghc-so-tightly-someone-turns-blue projects to always need to use `ghc-internals` too. I think it's less important to deny that category exists, than emphasize that the vast majority of projects are *not* in that category.

Put another way, I'd expect those projects to know who they are and not reach out to GHC devs or the CLC, and I'd expect almost everyone who does reach out to be told "actually, you don't need to do things that way". Is that what you were thinking too? Or were you imagining we might sometimes move something to `ghc-internals` (without a reexport elsewhere) that shouldn't be there, by mistake?

Reviewer 8: If such projects are relying on functionality from `base` that GHC developers and the CLC both agree should be moved to `ghc-internals`, they ought to reach out and say "hey, we use this for such-and-such", and make a case for it to be put in `ghc-prim` instead.

The point here is to make, FAIAP, a "hidden" library containing code in the strictest intersection of "compiler bootstrap" and "necessary to allow general-purpose programming": stuff like the guts of IO, datatype representations, and the like. There should be nothing  in `ghc-internals` that is both:
- useful for general programming, and
- not exported from `base`, `ghc-prim`, or `ghc-experimental`.

Author: I agree with the main "There should be nothing in ghc-internals that is both: useful for general programming, and not exported from..." point, but do not that I don't think anyone intends `ghc-prim` to be any more user-facing than `ghc-internals`. Perhaps the proposal should make this more explicit?

Reviewer 2: > not exported from base, ghc-prim, or ghc-experimental.

This should be "not exported fomr base or ghc-experimental".  `ghc-prim` is every bit "internal" as `ghc-internals`.   Anything in `ghc-prim` that is useful for general programming should be be exposed by `base` or `ghc-experimenal`.  Indeed maybe we should make that clearer.

  So we may need to move code from ``base`` to ``ghc-internals``, leaving a shim behind in base.
  In practice, that may mean that quite a lot of code will move into ``ghc-internals`` quite quickly.
  But that's fine: *it is just an implementation matter*: provided the modules, exports, and API of ``base`` are maintained, it is immaterial to clients (and hence to CLC) exactly *how* they are maintained.

- This proposal is fully compatible with, and actively supports, the `CLC charter <https://github.com/haskell/core-libraries-committee#base-package>`__:

    The primary responsibility of CLC is to manage API changes of ``base`` package.
    The ownership of ``base`` belongs to GHC developers, and they can maintain it freely without CLC involvement as long as changes are invisible to clients.
    Documentation changes normally fall under GHC developers purview, except significant ones (e.g., adding or changing type class laws).

- It also supports GHC innovation, by

  - allowing GHC freedom to change aspects of its implementation

  - allowing the GHC Steering Committee to add new functions and types in ``ghc-experimental``.

> - One might wonder why GHC has three "internal" packages: ``ghc-internals``, ``ghc-bignum``, and ``ghc-prim``? Could they not be a single package? Answer: technically yes, but it helps to keep dependencies and responsibilities clear.
>   And it's purely an internal GHC matter; if the team wants to structure GHC's internals with three packages, or ten, that's up to them.

Reviewer 8: Let's change this awkward phrasing:

```suggestion
- The three "internal" packages: ``ghc-internals``, ``ghc-bignum``, and ``ghc-prim``, could arguably be a single package, but the GHC team has decided that encapsulating the relevant code in this way helps to keep dependencies and responsibilities clear.
```

Author: I am fond of the second sentence, because its perhapps the only indication that `ghc-internals`'s very existence is also unstable; this follows from the GHC devs being allows to shuffle around code amongst their unstable libraries as often as they please.

Reviewer 2: @mixphix's rephrasing of the first sentence is helpful. (I would remove "arguably" or replace with "in principle".)  But as John says, let's retain "It's purely an internal GHC matter"

Author: ```suggestion
- The three "internal" packages: ``ghc-internals``, ``ghc-bignum``, and ``ghc-prim``, could arguably be a single package, but the GHC team has decided that encapsulating the relevant code in this way helps to keep dependencies and responsibilities clear.
  And it's purely an internal GHC matter; if the team wants to structure GHC's internals with three packages, or ten, that's up to them.
```

Continuous integration
======================

A major difficulty is **knowing when the API of 'base' (as defined in Section 2) has changed.** A change requires CLC approval; but how do we know what commits (to ``base``, to ``ghc-internals``, to ``ghc-prim``) make such a change?

In the past we have relied on best efforts; but with a bunch of volunteers, mistakes will be made.
And mistakes can lead to a loss of trust.

The solution is obvious: we need to automate.
We therefore propose the following, as part of CI:

1. Compile a good chunk of Hackage (around 500 packages) against base.
   We already do this, and it is a huge help in reassuring ourselves that a change does not lead to accidental breakage.

2. Test if any of the types (incl their kinds), functions (incl their types) and instances exposed by the ``base`` API are accidentally changed by a commit.

>    This is definitely going to happen, soon: @bgamari already has a prototype.

Reviewer 3: This tool would be *amazing* to have for other libraries - I'd love to have something here so I could know for sure if a new release is versioned appropriately.

Reviewer: I think this already exists as https://kowainik.github.io/posts/policeman-bristol - or is what you're looking for something else?

Reviewer 9: `policeman` is a good approximation but when I examined it for use in GHC I found it had a few shortcomings:

 * it makes no attempt at comparing anything beyond the names of exports; that is, the types of bindings, exposed instances, `MINIMAL` declarations of classes,  and the like are not accounted for in the PVP assessment
 * there isn't a clear way to handle packages with platform-dependent exports; `base` is (unfortunately) such a package
 * it relies on `.hie` files, which we currently don't produce in the GHC build

- Approval of the proposal (by the GHC Steering Committee, with the non-binding vote of CLC) is not a guarantee that the final implementation will land;
  that depends on the implementation being well engineered etc (GHC team);
  and the implementor should make an explicit proposal to the CLC specifying the precise changes.

Abstraction leakage
-------------------

We may foresee a couple of ways in which changes in ``ghc-internals`` could become client visible:

- Occasionally, an error message may mention a fully qualified name for an out-of-scope identifier.
  For example (GHC test ``mod153``)::

    Ambiguous occurrence ‘id’
               It could refer to either ‘Prelude.id’,
                            imported from ‘Prelude’ at mod153.hs:2:8
                            (and originally defined in ‘GHC.Base’)
                         or ‘M.id’, defined at mod153.hs:2:21

  The "originally defined in" mentions a module; and if that module is in a package that is not imported, GHC will package-qualify the module name.
  And seeing ``ghc-internals:GHC.Base`` is perhaps less nice.
  This is not a new problem: we already package-qualify modules in ``ghc-prim``.

>   One solution is to remove the "originally defined in.." parenthesis for types and functions that would require such package qualification.

Reviewer 3: IMO a lot of this would be helped by tracking the import/re-export provenance more directly. "Imported from A and originally defined in E" is nice but it's also often very nice to know the exact chain of imports and re-exports that brought something into scope. GHC could then have a flag to trim this provenance information to the direct dependencies of a package. That way, if you find yourself depending on `ghc-internals:SecretType`, you can see exactly how it got introduced into your codebase.

Author: In asking https://github.com/ghc-proposals/ghc-proposals/pull/595#issuecomment-1607844684 I was wondering something similar --- ideally we we do now know the entire provenance of imports and (re)exports, and there are a number of things we can do with it.

- Another form of leakage could be: a new class in ``ghc-internals``, *not exposed in base*, that is given instances for existing data types.
  There is a risk that those instances might confusingly be visible to clients of ``base``.

>   If so, the CLC should at least be consulted.

Reviewer 3: Wouldn't you need to import the relevant class to have these instances be visible? And if it's only exposed in `ghc-internals`, wouldn't that require an explicit import of the class?

This may just be an artifact of how [instance visibility works across component boundaries](https://gitlab.haskell.org/ghc/ghc/-/issues/20529), though.

Reviewer 9: Not necessarily. The primary way that this exposure might happen is via Haddocks. For instance, imagine that we have a class, exposed via `ghc-experimental` (and not in `base`) but defined in `ghc-internals` which some class exposed via `base` is an instance of. Under `haddock`'s current logic, this instance would be shown in the documentation for `base`, despite the class itself not being visible in `base`.

Versions and backports
----------------------

We agree that the version number of ``ghc-internals`` may have a major bump between minor releases of GHC.
(Why? Because to fix the bug we change something in ``ghc-internals``.)

This makes an exception to a general rule: generally, a minor release of GHC (say 9.6.4) which only fixes bugs, never makes a major version bump to ``base``, or indeed any boot package.

We should discuss this (rather important) exception with the Stackage curators.

But this same issue could in principle affect ``base`` too.
Very occasionally a **bug-fix** might involve a change to the user-visible API.
Example: `role annotations on SNat <https://github.com/haskell/core-libraries-committee/issues/170>`__ (although there is a debate as to whether this specific change constitutes a "breaking change" under the PVP).

Under these circumstances we (together) will have to decide whether to

- Back port the fix, and not bump the major version of ``base`` (i.e. bend the PVP), or
- Bump the major version of base, but therefore be unable to fix the bug in the released GHC.

This is a decision for the CLC.

> See PVP issue https://github.com/haskell/pvp/issues/10.

Reviewer 3: This is a technical limitation of the way that `base` and GHC work, and would in principle be fixed by allow a GHC version to work with multiple versions of `base`, right?

Author: Are you proposing (eventually) to continue to ship the `base` with the bug and a new one with the bug? That would work. But I think the point in this case was that some bugs are not a matter of implementation but inherent to the interface itself.

(I remain very pro multiple `base` versions with one GHC, to be clear :).)

Reviewer 3: I think I'm confused - can a released GHC use a different version of `base`? I was under the impression that the `base <-> GHC` relationship was fixed, and every version of GHC can only use a single version of `base`. That means that GHC X.Y.Z must always use the same `base` version, and a new `base` version means a new release of GHC.

Author: I think there is much confusion around this topic because people confuse GHC the software with GHC the official bindist.

The answer is Yes, and it's not even a new feature. See https://gitlab.haskell.org/ghc/ghc/-/blob/master/compiler/GHC/Unit/Types.hs#L562.

The short version that is one is liable to shoot themselves in the foot if they have more than one `base` in a package database, but one is always free to have multiple package databases they will not mix together. In each "parallel universe" of the separate package database, there is only one `base`, and no potential for confusion.

GHC in fact has *no way of knowing* whether this is the "official base" that is was shipped with the bindist, or ones own. How could it now? (It isn't shipped with trusted ABI hashes for `base`, for example, on the contrary, it expects `base` to have *no* ABI hashes; it is maximally trusting of whatever `base` you give it!)

For a variety of reasons, no one has yet proposed we take advantage of this. But we can. There is no technical limitation, just the ergonomic/human problem of it being easy to get our package databases with their separate `base`s mixed up.

(I believe @alt-romes might be working on getting all wired-ing packages ABI hashes to solve the humans-getting-confused problem, but I am not sure about that.)

Reviewer 2: > (I remain very pro multiple base versions with one GHC, to be clear :).)

I think we all are in favour of that.   

But, to be clear, achieving that goal is not part of the current document.  So this thread is somewhat "by the way".

Still, the plan outlined this document should make the goal of a reinstallable base much more achievable.  GHC knows (wired into its binary) where, say `GHC.Base.map` is defined.  If we move that definition from one module to another, everything will stop working.  I think that is one reason that `base` and GHC are so tightly bound together.  But if  `GHC.Base.map` was in `ghc-internals` then GHC and `ghc-internals` would remain tightly bound together, but `base` would be much more loosely coupled.  I don't yet see any technical obstacle to installing a fresh `base` without changing `ghc`.

I could be wrong; but all that matters for *this* conversation is that the plan here is a step in the right direction for the reinstallable-base goal.

Reviewer 3: Fantastic, thanks for the clarification!

Allowing `base` to iterate without incurring a GHC release cycle would be a big win for the ecosystem, for sure.

All three packages conform rigorously to the PVP.
(But see Section 5.3)

Some observations about this structure:

- We should use all possible social and technical means to discourage people from depending directly on ``ghc-internals``, because if such dependencies become frequent and ossified, it will lead to future pain when the API changes.
  Saying "we told you not to rely on it" may be true, but won't lessen that pain.
  What mechanisms could we use?

  - The name ``ghc-internals`` is a pretty strong signal all by itself.

  - Cabal description and README explains how it is intended used (and not used).

  - Hoogle could (by default anyway) never show stuff from ``ghc-internals``.

  - Do not upload Haddocks for ``ghc-internals`` to Hackage.
    (Ditto ``ghc-prim``.) Need to make sure that if someone wants to follow the Haddock source-code link to (say) Functor, they should still find it regardless of where it is actually defined.

  - We could consider issuing a warning if you say ``-package ghc-internals`` (or ``ghc-bignum`` or ``ghc-prim``), one that was hard to silence.

>     Since we can have module-level ``WARNING`` pragmas with custom categories, one way to realise this would be to pick a category and add such pragmas to every module in the relevant packages, though we might want to do something more systematic.

Reviewer 4: This sounds excellent. Much like "partial", "internal" seems like a fairly natural warning category that other people in the ecosystem might want to use. And it's easy for a user to say "I know what I'm doing" by just turning off that warning category in the module where they use internals.

Reviewer 6: There's the `category` field in cabal files which is pretty unstructured an all over the map. We could easily initiate a convention of the `internal` category actually meaning something -- e.g. triggering an extra textual notice on hackage pages listing the package, and extending cabal so it could maybe optionally filter or warn on such stuff.

Reviewer 4: That seems like more work for less fine control? If we already have the mechanism in GHC at a module level, that seems sufficient to me.

Reviewer 6: Hackage itself cannot and should not read module-level pragmas, nor can cabal. This proposal is about having some _packages_ be internal and some not. So we need package-level mechanisms for marking things internal, such that both hackage and cabal can take appropriate action (perhaps additional html, like a big red warning box on hackage, and for cabal the ability to warn on using internals as a _direct_ rather than indirect dependency, etc)

Reviewer 5: > and for cabal the ability to warn on using internals as a direct rather than indirect dependency, etc

From what I understood people are also worried about indirect dependencies (coming from things other than the original package), because they may easily lack behind upgrades and bubble up churn.

Reviewer 6: Yes, that would be another thing that would be nice to address. the `category` field does not address that, but seems useful nonetheless, and I would encourage its consideration regardless.

Reviewer 4: (Sorry, I realised that what I was saying was ambiguous based on the line I was commenting on: I'm in favour of a module-level INTERNAL warning, _added manually to source files_, not magically added to packages by cabal or Hackage)
